src/lambdaisland/witchcraft/citizens/trait.clj

```diff
@@ -0,0 +1,123 @@
+  desired namespaces just like any other class. See gen-class for a
+  description of the options.
+
+  Taken from the rich comment section of `clojure/genclass.clj`.
+  "
+  [options-map]
+  (let [[cname bytecode] (#'clojure.core/generate-class options-map)]
+    (.defineClass (licp/root-loader)
+                  (str/replace cname
+                               #"/"
+                               ".")
+                  bytecode
+                  "")))
+
+(defonce
+  ^{:doc "Registry from trait name to mape of callbacks"}
+  registry
+  (atom {}))
+
+(defn- call-cb
+  "Invoke a callback for a given trait and callback name. Callback receives
+  `this` (the trait) and any additional args."
+  [this cb-name & args]
+  (when-let [cb (get-in @registry [(.getName this) cb-name])]
+    (apply cb this args)))
+
+;; The callbacks we understand
+(defn trait-load [this key] (call-cb this :load key))
+(defn trait-onAttach [this] (call-cb this :on-attach))
+(defn trait-onCopy [this] (call-cb this :on-copy))
+(defn trait-onDespawn [this] (call-cb this :on-despawn))
+(defn trait-onPreSpawn [this] (call-cb this :on-pre-spawn))
+(defn trait-onSpawn [this] (call-cb this :on-spawn))
+(defn trait-run [this] (call-cb this :run))
+(defn trait-save [this key] (call-cb this :save key))
+
+(defn trait-isRunImplemented [this]
+  (boolean (get-in @registry [(.getName this) :run])))
+
+(defn make-trait
+  "Create a subclass of npcs.citizensapi.trait.Trait, `name` is a simple lowercase
+  identifier of the trait like `\"sneak\"` or `\"aggressive\"`. `callbacks` is a
+  map from keyword to function (or var, IFn). Callbacks receive `this` (the
+  trait) as first arg. `:load` and `:save` also receive a `DataKey`.
+
+  Callbacks: `:load`, `:save`, `:on-attach`, `:on-copy`, `:on-despawn`,
+  `:on-pre-spawn`, `:run`, `:save`.
+
+  Can be called multiple times, subsequent calls will only replace the
+  callbacks. Returns the trait subclass `java.lang.Class`.
+
+  Note that adding a `:run` callback for a Trait that is already attached to an
+  NPC will not cause it to be called. Remove the Trait from the NPC and add it
+  back.
+  "
+  [name callbacks]
+  (let [name (str/lower-case name) ; Trait does the same, stay consistent
+        init-fn-name (str "init-" name)
+        classname (str *ns* "." name)]
+    (swap! registry assoc name callbacks)
+    ;; Make a unique `init` function for this class, so we have a way to inject
+    ;; the correct name into the superclass constructor
+    (intern *ns* (symbol (str "trait-" init-fn-name)) (fn [] [[name]]))
+    (try
+      (Class/forName classname)
+      (catch java.lang.ClassNotFoundException _
+        (gen-and-load-class
+         {:name classname
+          :extends Trait
+          ;; :state "state"
+          :init init-fn-name
+          :constructors {[] [String]}
+          ;;:post-init `my-post-init
+          :prefix "trait-"
+          ;;:impl-ns "my.trait.impl"
```

Comment:
All of our trait subclasses will share the same method implementations, but these implementations simply delegate to the trait-specific callbacks in the registry. I do it this way because reloading classes/class methods when they change is not possible (or at least a huge pain). The extra indirection gives us the possibility to easily redefine things.

Reply:
I see, that is why you go via `call-cb` and `registry`. Thanks!